File: native-android/app/src/main/kotlin/com/crossdashboard/app/data/repository/MemoRepository.kt
package com.crossdashboard.app.data.repository

import android.util.Log
import com.crossdashboard.app.data.db.Mappers
import com.crossdashboard.app.data.db.dao.MemosDao
import com.crossdashboard.app.data.db.toDomain
import com.crossdashboard.app.data.db.toEntity
import com.crossdashboard.app.data.network.MemosClient
import com.crossdashboard.app.data.prefs.CredentialKey
import com.crossdashboard.app.data.prefs.SecureStore
import com.crossdashboard.app.domain.model.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MemoRepository @Inject constructor(
    private val dao: MemosDao,
    private val client: MemosClient,
    private val secureStore: SecureStore,
) {
    private val tag = "MemoRepository"

    // ─── Observe ──────────────────────────────────────────────────────────────

    val allMemos: Flow<List<MemosMemo>> = dao.observeAll().map { list ->
        list.map { it.toDomain() }
    }

    val normalMemos: Flow<List<MemosMemo>> = dao.observeNormal().map { list ->
        list.map { it.toDomain() }
    }

    val archivedMemos: Flow<List<MemosMemo>> = dao.observeArchived().map { list ->
        list.map { it.toDomain() }
    }

    // ─── Sync ─────────────────────────────────────────────────────────────────

    /** Pages through all memos (NORMAL + ARCHIVED) and replaces the local cache. */
    suspend fun syncMemos() = withContext(Dispatchers.IO) {
        if (client.baseUrl() == null) return@withContext
        val allFetched = mutableListOf<MemosMemo>()

        for (state in listOf(MemoState.NORMAL, MemoState.ARCHIVED)) {
            var pageToken: String? = null
            do {
                val (page, next) = client.listMemos(pageToken = pageToken, state = state)
                allFetched += page
                pageToken = next
            } while (pageToken != null)
        }

        if (allFetched.isNotEmpty()) {
            runCatching {
                dao.deleteAll()
                dao.upsertAll(allFetched.map { it.toEntity() })
            }.onFailure { e -> Log.e(tag, "syncMemos DB write failed", e) }
        }
    }

    // ─── Create ───────────────────────────────────────────────────────────────

    /**
     * Uploads each [PendingAttachment] first, then creates the memo with the
     * returned attachment names.
     */
    suspend fun createMemo(
        content: String,
        visibility: MemoVisibility,
        attachments: List<PendingAttachment>,
    ): MemosMemo? = withContext(Dispatchers.IO) {
        val attachmentNames = attachments.mapNotNull { pending ->
            client.createAttachment(
                filename = pending.fileName,
                mimeType = pending.mimeType,
                bytes = pending.bytes,
            )?.name
        }
        val memo = client.createMemo(
            content = content,
            visibility = visibility,
            attachmentNames = attachmentNames,
        ) ?: return@withContext null
        runCatching { dao.upsert(memo.toEntity()) }
            .onFailure { e -> Log.e(tag, "createMemo DB insert failed", e) }
        memo
    }

    // ─── Delete ───────────────────────────────────────────────────────────────

    suspend fun deleteMemo(name: String, force: Boolean = false) = withContext(Dispatchers.IO) {
        client.deleteMemo(name, force)
        runCatching { dao.deleteByName(name) }
            .onFailure { e -> Log.e(tag, "deleteMemo DB remove failed", e) }
    }

    // ─── Archive ──────────────────────────────────────────────────────────────

    suspend fun archiveMemo(name: String): MemosMemo? = withContext(Dispatchers.IO) {
        val updated = client.updateMemo(
            memoId = name,
            state = MemoState.ARCHIVED,
            updateMask = "state",
        ) ?: return@withContext null
        runCatching { dao.upsert(updated.toEntity()) }
            .onFailure { e -> Log.e(tag, "archiveMemo DB update failed", e) }
        updated
    }

    suspend fun restoreMemo(name: String): MemosMemo? = withContext(Dispatchers.IO) {
        val updated = client.updateMemo(
            memoId = name,
            state = MemoState.NORMAL,
            updateMask = "state",
        ) ?: return@withContext null
        runCatching { dao.upsert(updated.toEntity()) }
            .onFailure { e -> Log.e(tag, "restoreMemo DB update failed", e) }
        updated
    }

    // ─── Comments ─────────────────────────────────────────────────────────────

    /** Live network call — comments are not cached locally. */
    suspend fun loadComments(memoId: String): List<MemosMemo> = withContext(Dispatchers.IO) {
        client.listMemoComments(memoId)
    }

    suspend fun createComment(parentId: String, content: String): MemosMemo? =
        withContext(Dispatchers.IO) {
            client.createMemoComment(parentId, content)
        }

    // ─── Share ────────────────────────────────────────────────────────────────

    /** Returns the full share URL or null on error. */
    suspend fun createShare(memoId: String): String? = withContext(Dispatchers.IO) {
        client.createMemoShare(memoId)
    }

    // ─── Share resolution ─────────────────────────────────────────────────────

    suspend fun getMemoByShare(shareId: String): MemosMemo? = withContext(Dispatchers.IO) {
        client.getMemoByShare(shareId)
    }
}

/** Pending file to attach — mirrors GiteaClient PendingAttachment pattern from IssuesViewModel. */
data class PendingAttachment(
    val fileName: String,
    val mimeType: String,
    val bytes: ByteArray,
) {
    override fun equals(other: Any?): Boolean =
        other is PendingAttachment && fileName == other.fileName && mimeType == other.mimeType

    override fun hashCode(): Int = 31 * fileName.hashCode() + mimeType.hashCode()
}

File: native-macos/CrossDashboard/Data/Repository/MemoRepository.swift
import Foundation
import SwiftData
import CrossDashboardKit

/// Mirrors MemoRepository.kt.
@Observable
@MainActor
final class MemoRepository {

    private let context: ModelContext
    private let client: MemosClient

    var allMemos: [MemosMemo] = []
    var normalMemos: [MemosMemo] { allMemos.filter { $0.state == .normal } }
    var archivedMemos: [MemosMemo] { allMemos.filter { $0.state == .archived } }

    init(context: ModelContext, client: MemosClient) {
        self.context = context
        self.client  = client
        loadFromDB()
    }

    // ─── DB ───────────────────────────────────────────────────────────────────

    func loadFromDB() {
        var descriptor = FetchDescriptor<MemosModel>()
        descriptor.includePendingChanges = false
        let models = (try? context.fetch(descriptor)) ?? []
        allMemos = models.map { $0.toDomain() }.sorted { $0.displayTime > $1.displayTime }
    }

    // ─── Sync ─────────────────────────────────────────────────────────────────

    @discardableResult
    func syncMemos() async -> Bool {
        guard client.baseUrl() != nil else { return true }
        var fetched: [MemosMemo] = []

        for state in [MemoState.normal, .archived] {
            var pageToken: String? = nil
            repeat {
                guard let result = await client.listMemosForSync(
                    pageToken: pageToken,
                    state: state
                ) else { return false }
                fetched.append(contentsOf: result.memos)
                pageToken = result.nextPageToken
            } while pageToken != nil
        }

        do {
            try context.transaction {
                try context.delete(model: MemosModel.self)
                fetched.forEach { context.insert(MemosModel(from: $0)) }
            }
        } catch {
            context.rollback()
            return false
        }
        loadFromDB()
        return true
    }

    // ─── Create ───────────────────────────────────────────────────────────────

    func createMemo(
        content: String,
        visibility: MemoVisibility,
        attachments: [MemoPendingAttachment]
    ) async -> MemosMemo? {
        var names: [String] = []
        for att in attachments {
            if let uploaded = await client.createAttachment(
                filename: att.filename,
                mimeType: att.mimeType,
                data: att.data
            ) {
                names.append(uploaded.name)
            }
        }
        guard let memo = await client.createMemo(
            content: content,
            visibility: visibility,
            attachmentNames: names
        ) else { return nil }
        context.insert(MemosModel(from: memo))
        try? context.save()
        loadFromDB()
        return memo
    }

    // ─── Delete ───────────────────────────────────────────────────────────────

    func deleteMemo(_ name: String, force: Bool = false) async {
        await client.deleteMemo(name, force: force)
        let models = (try? context.fetch(FetchDescriptor<MemosModel>())) ?? []
        models.filter { $0.name == name }.forEach { context.delete($0) }
        try? context.save()
        loadFromDB()
    }

    // ─── Archive / Restore ───────────────────────────────────────────────────

    func archiveMemo(_ name: String) async -> MemosMemo? {
        guard let updated = await client.updateMemo(name, state: .archived, updateMask: "state")
        else { return nil }
        upsertInDB(updated)
        return updated
    }

    func restoreMemo(_ name: String) async -> MemosMemo? {
        guard let updated = await client.updateMemo(name, state: .normal, updateMask: "state")
        else { return nil }
        upsertInDB(updated)
        return updated
    }

    // ─── Comments ─────────────────────────────────────────────────────────────

    func loadComments(_ memoId: String) async -> [MemosMemo] {
        await client.listMemoComments(memoId)
    }

    func createComment(parentId: String, content: String) async -> MemosMemo? {
        await client.createMemoComment(parentMemoId: parentId, content: content)
    }

    // ─── Share ────────────────────────────────────────────────────────────────

    func createShare(_ memoId: String) async -> String? {
        await client.createMemoShare(memoId)
    }

    func getMemoByShare(_ shareId: String) async -> MemosMemo? {
        await client.getMemoByShare(shareId)
    }

    // ─── Relations ────────────────────────────────────────────────────────────

    func listRelations(_ memoId: String) async -> [MemoRelationSwift] {
        await client.listMemoRelations(memoId)
    }

    // ─── Private ──────────────────────────────────────────────────────────────

    private func upsertInDB(_ memo: MemosMemo) {
        let models = (try? context.fetch(FetchDescriptor<MemosModel>())) ?? []
        if let existing = models.first(where: { $0.name == memo.name }) {
            context.delete(existing)
        }
        context.insert(MemosModel(from: memo))
        try? context.save()
        loadFromDB()
    }
}

// ─── Pending attachment for UI layer ─────────────────────────────────────────

struct MemoPendingAttachment: Sendable {
    let filename: String
    let mimeType: String
    let data: Data
}
